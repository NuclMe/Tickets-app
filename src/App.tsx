import Box from '@mui/material/Box';
import type { FC } from 'react';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Routes, Route, Navigate } from 'react-router-dom';

import { Header, Transfer, Flights, Items } from './components';

import { GlobalStyles } from './GlobalStyles';
import {
  fetchTickets,
  selectFilteredTickets,
} from './redux/ducks/tickets/tickets';
import { MainContentWrapper, RootContainer } from './styledVars';
import type { ITicket } from './types';

export const App: FC = () => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchTickets());
  }, [dispatch]);

  const filteredTickets = useSelector(selectFilteredTickets);

  const cheapestTickets: ITicket[] = [...filteredTickets].sort(
    (a, b) => a.price - b.price
  );
  const fastestTickets: ITicket[] = [...filteredTickets].sort(
    (a, b) =>
      a.segments[0].duration +
      a.segments[1].duration -
      (b.segments[0].duration + b.segments[1].duration)
  );

  return (
    <RootContainer>
      <GlobalStyles />
      <Header />
      <MainContentWrapper>
        <Transfer />
        <Box sx={{ display: 'flex', flexDirection: 'column', flex: 3 }}>
          <Flights />
          <Routes>
            <Route path="/" element={<Navigate to="/cheapest" replace />} />
            <Route
              path="/cheapest"
              element={<Items items={cheapestTickets} />}
            />
            <Route path="/fastest" element={<Items items={fastestTickets} />} />
          </Routes>
        </Box>
      </MainContentWrapper>
    </RootContainer>
  );
};
