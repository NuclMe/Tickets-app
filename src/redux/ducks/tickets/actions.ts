import { types } from './types';

export const fetchTickets = () => ({ type: types.FETCH_TICKETS });
export const fetchTicketsSuccess = (tickets) => ({
  type: types.FETCH_TICKETS_SUCCESS,
  payload: tickets,
});
export const fetchTicketsError = (error) => ({
  type: types.FETCH_TICKETS_ERROR,
  error,
});

export const setTransferFilter = (filter) => ({
  type: types.SET_TRANSFER_FILTER,
  payload: filter,
});
