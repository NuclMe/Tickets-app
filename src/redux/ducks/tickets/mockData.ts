import type { ITicket } from '../../../types';

export const mockTickets: ITicket[] = [
  {
    price: 35230,
    carrier: 'SU',
    segments: [
      {
        origin: 'MOW',
        destination: 'HKT',
        date: '2018-05-18T02:00:00.000Z',
        stops: [],
        duration: 400,
      },
      {
        origin: 'HKT',
        destination: 'MOW',
        date: '2018-05-18T08:30:00.000Z',
        stops: [],
        duration: 300,
      },
    ],
  },
  {
    price: 30000,
    carrier: 'S7',
    segments: [
      {
        origin: 'MOW',
        destination: 'HKT',
        date: '2018-05-18T02:00:00.000Z',
        stops: ['LED'],
        duration: 600,
      },
      {
        origin: 'HKT',
        destination: 'MOW',
        date: '2018-05-18T08:30:00.000Z',
        stops: ['LED'],
        duration: 200,
      },
    ],
  },
  {
    price: 25000,
    carrier: 'TK',
    segments: [
      {
        origin: 'MOW',
        destination: 'HKT',
        date: '2018-05-18T02:00:00.000Z',
        stops: ['IST', 'DXB'],
        duration: 750,
      },
      {
        origin: 'HKT',
        destination: 'MOW',
        date: '2018-05-18T08:30:00.000Z',
        stops: ['IST', 'DXB'],
        duration: 650,
      },
    ],
  },
  {
    price: 40000,
    carrier: 'BA',
    segments: [
      {
        origin: 'MOW',
        destination: 'HKT',
        date: '2018-05-18T02:00:00.000Z',
        stops: ['LED'],
        duration: 500,
      },
      {
        origin: 'HKT',
        destination: 'MOW',
        date: '2018-05-18T08:30:00.000Z',
        stops: ['LED', 'AMS'],
        duration: 450,
      },
    ],
  },
  {
    price: 45000,
    carrier: 'LH',
    segments: [
      {
        origin: 'MOW',
        destination: 'HKT',
        date: '2018-05-18T02:00:00.000Z',
        stops: ['FRA', 'BKK', 'SIN'],
        duration: 800,
      },
      {
        origin: 'HKT',
        destination: 'MOW',
        date: '2018-05-18T08:30:00.000Z',
        stops: ['SIN', 'BKK', 'FRA'],
        duration: 750,
      },
    ],
  },
];
