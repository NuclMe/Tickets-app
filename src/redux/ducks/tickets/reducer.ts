import { types } from './types';
const initialState = {
  tickets: [],
  loading: false,
  error: null,
  selectedTransfer: 'all',
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case types.FETCH_TICKETS:
      return { ...state, loading: true, error: null };
    case types.FETCH_TICKETS_SUCCESS:
      return { ...state, loading: false, tickets: action.payload };
    case types.FETCH_TICKETS_ERROR:
      return { ...state, loading: false, error: action.error };
    case types.SET_TRANSFER_FILTER:
      return { ...state, selectedTransfer: action.payload };
    default:
      return state;
  }
};
