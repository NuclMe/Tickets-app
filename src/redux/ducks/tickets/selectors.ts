import { createSelector } from 'reselect';

const ticketsSelector = (state) => state.ticket.tickets;
const transferSelector = (state) => state.ticket.selectedTransfer;

export const selectFilteredTickets = createSelector(
  [ticketsSelector, transferSelector],
  (tickets, selectedTransfer) => {
    if (selectedTransfer === 'all') return tickets;
    return tickets.filter((ticket) =>
      ticket.segments.every((seg) => seg.stops.length === selectedTransfer)
    );
  }
);
