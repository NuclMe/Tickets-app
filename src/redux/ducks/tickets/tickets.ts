import { createSelector } from 'reselect';
import type { RootState } from '../../root-reducer';
import { types } from './types';
import type { ITicket } from '../../../types';
import { call, put, takeLatest } from 'redux-saga/effects';
import axios from 'axios';

const initialState = {
  tickets: [],
  loading: false,
  error: null,
  selectedTransfer: 'all',
};

const ticketsSelector = (state: RootState) => state.tickets.tickets;
const transferSelector = (state: RootState) => state.tickets.selectedTransfer;

export const selectFilteredTickets = createSelector(
  [ticketsSelector, transferSelector],
  (tickets: ITicket[], selectedTransfer: string) => {
    if (selectedTransfer === 'all') return tickets;
    return tickets.filter((ticket: ITicket) =>
      ticket.segments.every(
        (seg) => seg.stops.length === Number(selectedTransfer)
      )
    );
  }
);

export const fetchTickets = () => ({ type: types.FETCH_TICKETS });
export const fetchTicketsSuccess = (tickets: ITicket[]) => ({
  type: types.FETCH_TICKETS_SUCCESS,
  payload: tickets,
});
export const fetchTicketsError = (error: string) => ({
  type: types.FETCH_TICKETS_ERROR,
  error,
});
export const setTransferFilter = (filter: number | string) => ({
  type: types.SET_TRANSFER_FILTER,
  payload: filter,
});

export const reducer = (state = initialState, action: any) => {
  switch (action.type) {
    case types.FETCH_TICKETS:
      return { ...state, loading: true, error: null };
    case types.FETCH_TICKETS_SUCCESS:
      return { ...state, loading: false, tickets: action.payload };
    case types.FETCH_TICKETS_ERROR:
      return { ...state, loading: false, error: action.error };
    case types.SET_TRANSFER_FILTER:
      return { ...state, selectedTransfer: action.payload };
    default:
      return state;
  }
};
function* fetchTicketsSaga(): Generator {
  try {
    let allTickets: ITicket[] = [];
    let stop = false;
    let retries = 0;
    const maxRetries = 10;

    while (!stop && retries < maxRetries) {
      try {
        const ticketsRes = yield call(axios.get, '/mock-api/tickets.json');
        allTickets = allTickets.concat(ticketsRes.data.tickets);
        stop = ticketsRes.data.stop;
      } catch (err) {
        console.error(err);
        retries += 1;
        yield call(() => new Promise((res) => setTimeout(res, 500)));
        continue;
      }
    }

    yield put(fetchTicketsSuccess(allTickets));
  } catch (error: any) {
    yield put(fetchTicketsError(error.message));
  }
}

export function* ticketSaga() {
  yield takeLatest(types.FETCH_TICKETS, fetchTicketsSaga);
}
