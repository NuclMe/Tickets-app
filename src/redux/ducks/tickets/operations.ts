import { call, put, takeLatest } from 'redux-saga/effects';
import axios from 'axios';
import { types } from './types';
import { fetchTicketsSuccess, fetchTicketsError } from './actions';

function* fetchTicketsSaga() {
  try {
    let allTickets = [];
    let stop = false;
    let retries = 0;
    const maxRetries = 10;

    while (!stop && retries < maxRetries) {
      try {
        const ticketsRes = yield call(axios.get, '/mock-api/tickets.json');
        allTickets = allTickets.concat(ticketsRes.data.tickets);
        stop = ticketsRes.data.stop;
      } catch (err) {
        retries += 1;
        yield call(() => new Promise((res) => setTimeout(res, 500)));
        continue;
      }
    }

    yield put(fetchTicketsSuccess(allTickets));
  } catch (error) {
    yield put(fetchTicketsError(error.message));
  }
}

export function* ticketSaga() {
  yield takeLatest(types.FETCH_TICKETS, fetchTicketsSaga);
}
