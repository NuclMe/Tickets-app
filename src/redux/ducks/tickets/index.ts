import { fetchTickets } from './actions';
import { reducer } from './reducer';
import { ticketSaga } from './operations';
import { selectFilteredTickets } from './selectors';

export { fetchTickets, reducer, ticketSaga, selectFilteredTickets };
