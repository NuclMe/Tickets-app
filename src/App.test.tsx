import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { describe, expect, it } from 'vitest';

import { App } from './App';
import { reducer as ticketsReducer } from './redux/ducks/tickets/tickets';

const createTestStore = () => {
  return configureStore({
    reducer: {
      tickets: ticketsReducer,
    },
  });
};

const TestWrapper = ({ children }) => {
  const store = createTestStore();
  return (
    <Provider store={store}>
      <BrowserRouter>{children}</BrowserRouter>
    </Provider>
  );
};

describe('App', () => {
  it('renders the App component via snapshot', () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    expect(App).toMatchSnapshot();
  });
});
