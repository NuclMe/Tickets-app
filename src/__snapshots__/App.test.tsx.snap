// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders the App component via snapshot 1`] = `[Function]`;
